// performance_analysis/performance_analysis_es6.js
	//https://docs.npmjs.com/getting-started/creating-node-modules
	//https://googlechrome.github.io/samples/classes-es6/
	/*
		Search query must supply:
			* name		| Provides a label for the row
			* value		| Provides a measurement to represent the cell background color
			* status	| Provides a metric for the icon in the cell
			-------
			Optional values:
			* warning_threshold	| Level at which green becomes amber
			* critical_threshold	| Level at which amber becomes red	
	*/
	
	
	class performance_analysis{
		constructor(granularity, warning_threshold, critical_threshold,down_time_start,down_time_end,time_format,show_legend,showStatusAsText,OKIsLowOrHigh,sortOrder){
			this.items = [];
			this.num_items = 0;
			this.start_time = Date.now();
			this.end_time = Date.now();
			this.granularity = granularity;
			this.numBuckets = 12;
			this.warning_threshold = warning_threshold;
			this.critical_threshold = critical_threshold;
			this.OKIsLowOrHigh = (OKIsLowOrHigh.toLowerCase()=="low") ? "low" : "high";
			//Colour Defaults:
			this.okColour = "#78B24A";
			this.warningColour = "#E0C135";
			this.criticalColour = "#DD0000";
			this.noDataColour = "#5EBFC6";
			this.downTimeStart=down_time_start;
			this.downTimeEnd=down_time_end;
			this.timeFormat = time_format;
			this.showLegend = (show_legend=="true");
			this.showStatusAsText = (showStatusAsText=="true");
			this.sortOrder = sortOrder.toLowerCase();
			if (this.sortOrder!="az" && this.sortOrder!="za") {this.sortOrder="default";}
			//----------------
			this.data_item;
			this.current_item;
		}
		

		//Method to set all the colours for OK / Warning/ Critical / NoData
		set_colours(okColour, warningColour, criticalColour, noDataColour){
			this.okColour = okColour;
			this.warningColour = warningColour;
			this.criticalColour = criticalColour;
			this.noDataColour = noDataColour;
		}


		setData(data){
			var temp_bucket;
			var bucket;
			var bucket_start_time_seconds;
			var bucket_end_time_seconds;
			var bucket_index = 0;
			var start_time_in_seconds;
			var time_minute = 0;
			var time_hour = 0;
			var i = 0;
			var fields = [];
			var data_item;
			var current_item;
			var tempTime
			var k;
			var oMoment = require('moment');
			var date_first_row;
			var date_earliest = "-";
			var date_latest = "-";
			var date_last_row;
			var date_temp;
			var SplunkVisUtils = require('api/SplunkVisualizationUtils');
			var vizUtils = require('api/SplunkVisualizationUtils');
			const { performance_analysis, item, time_bucket } = require('performance_analysis');;
			
			try{
				//------------------------------  Get data row field indexes ----------------------------------------------------------------------
				for (i=0; i<data.fields.length; i++){
					fields[data.fields[i].name.toLowerCase()]  = i;
				}
				
				// Find earliest and latest timestamp
				for (i=0; i<data.rows.length; i++){
					tempTime = oMoment(data.rows[i][fields["_time"]]);
					if(date_earliest=="-") { date_earliest=tempTime;}
					if(date_latest=="-") { date_latest=tempTime;}
					if (tempTime > date_latest) {
						date_latest=tempTime
					} else if(tempTime < date_earliest) {
						date_earliest = tempTime;
					}
				}
					this.start_time=date_earliest;
					this.end_time = date_latest;
				
				//Snap to the granularity - e.g. if Granularity is 5 minutes, set start/end time to hh:00:00, hh:05:00, hh:10:00
				if(this.granularity<=60){
					time_minute = oMoment(this.start_time).minute()
					this.start_time=oMoment(this.start_time).minute(Math.floor(time_minute/this.granularity)* this.granularity).second(0).toDate();
					time_minute = oMoment(this.end_time).minute()
					this.end_time=oMoment(this.end_time).minute(Math.ceil(time_minute/this.granularity)* this.granularity).second(0).toDate();
				}else{
					// Granularity is more than 1 hour - just set start minutes/seconds to hh:00:00 and end minutes/seconds to hh:59:59
					this.start_time=oMoment(this.start_time).minute(0).second(0).toDate();
					this.end_time=oMoment(this.end_time).minute(59).second(59).toDate();
				}
				
				var currentTime = this.start_time.getTime();
				var localOffset = (-1) * this.start_time.getTimezoneOffset() * 60000;
				var start_time_in_seconds = Math.round(new Date(currentTime + localOffset).getTime() / 1000);
				var start_time_in_seconds_utc = this.start_time.getTime()/1000;
				// Timezone won't affect the number of buckets
				this.num_buckets = Math.ceil(((this.end_time.getTime() / 1000)  - (this.start_time.getTime() / 1000) ) / (this.granularity*60))+1;
				//---------------- End set start / end time from data object ----------------------------


				// Get the current transation, or create it if it doesn't yet exist
				for (i=0; i<data.rows.length; i++){
					data_item = data.rows[i];
					//------------------------------  Create or Locate item Objects ----------------------------------------------------------------------
					if(typeof this.items[vizUtils.escapeHtml(data_item[fields["name"]])] === 'undefined') {
						// does not exist
						this.items[vizUtils.escapeHtml(data_item[fields["name"]])] = new item(vizUtils.escapeHtml(data_item[fields["name"]]));
						this.num_items++;
						current_item = this.items[vizUtils.escapeHtml(data_item[fields["name"]])];
						current_item.num_buckets = this.num_buckets;
						// Create Time Buckets for this item
						for(k=0;k<current_item.num_buckets;k++){
							//Set up the time bucket
							bucket_start_time_seconds = start_time_in_seconds + (k *  this.granularity * 60);
							bucket_end_time_seconds = start_time_in_seconds + ((1+k) * this.granularity * 60) -1;
							temp_bucket = new time_bucket();
							temp_bucket.showStatusAsText = this.showStatusAsText;
							temp_bucket.timeFormat = this.timeFormat;
							temp_bucket.set_colours(this.okColour, this.warningColour, this.criticalColour, this.noDataColour);
							temp_bucket.warning_threshold = data_item[fields["warning_threshold"]] || this.warning_threshold;
							temp_bucket.critical_threshold = data_item[fields["critical_threshold"]] || this.critical_threshold;
							temp_bucket.set_start_time(bucket_start_time_seconds);
							temp_bucket.set_end_time(bucket_end_time_seconds);
							temp_bucket.set_category(current_item.name);
							temp_bucket.OKIsLowOrHigh = this.OKIsLowOrHigh; 
							// Is the bucket during down time?
							if(oMoment(temp_bucket.start_time).hour() >= this.downTimeStart && oMoment(temp_bucket.start_time).hour() < this.downTimeEnd) temp_bucket.isInDownTime = true;
							//Add to item's bucket array
							current_item.buckets.push(temp_bucket);
						}
					} else {
						// does exist
					}
					//------------------------------  Fill in details for Bucket Objects ----------------------------------------------------------------------
					current_item = this.items[vizUtils.escapeHtml(data_item[fields["name"]])];		
					// Find the correct bucket to put the item details into:
					// Index = Math.ceiling(   (time - start_time) / granularity   ) 
					bucket_start_time_seconds = oMoment(data_item[fields["_time"]]).toDate().getTime()/1000; 
					bucket_index = Math.ceil(   (bucket_start_time_seconds - (start_time_in_seconds_utc-1)) / (60*this.granularity)) -1;
					if (bucket_index >= current_item.buckets.length) bucket_index = current_item.buckets.length-1;
					current_item.buckets[bucket_index].add_item(data_item, fields);
				}	
			} catch(err) {
				console.log("Error setting data. " + err);
			}
		}

	
		getHTML(){
			var item_counter = 0;
			var bucket_counter = 0;
			var css="";
			var html = '<table class="performance_analysis">';
			var item;
			var oMoment = require('moment');
			var item_names;

			if(this.sortOrder =="default"){
				item_names = Object.keys(this.items);
			}else if(this.sortOrder == "az"){
				item_names = Object.keys(this.items).sort();
			}else{
				item_names = Object.keys(this.items).sort().reverse();
			}
			
			
			for(item_counter=0; item_counter < item_names.length ; item_counter++){
				item = this.items[item_names[item_counter]];
				
				// Print out the header row of the table
				if(item_counter==0){
					html += '<tr><td>&nbsp;</td>';
					for(bucket_counter=0; bucket_counter < this.num_buckets -1; bucket_counter++){
						//If time is in a downtime period, add the downtime class
						css = (item.buckets[bucket_counter].isInDownTime) ? " downtime" : "";
						html += '<td class="time' + css + '">' + oMoment(item.buckets[bucket_counter].start_time).format(this.timeFormat) + '</td>';
					}
					html+= '</tr>';
				}
				//---------------------------------------
				html += '<tr><tr><td class="item_name">' + item.name + '</td>';
				for(bucket_counter=0; bucket_counter < this.num_buckets -1; bucket_counter++){
					html += item.buckets[bucket_counter].getHTML();
				}
				html += '</tr>';
			} // loop item_counter
			html += '</table>';
			//Add a legend?
			if(this.showLegend){
				html += '<table class="performance_analysis_legend_container"><tr><td><a href="#" onclick="$(\'table.performance_analysis_legend\').toggle();return false;" title="View Legend"><img src="/static/app/performance-analysis/images/legend.gif" alt="Legend" /></a></td><td>';
				html += '<table class="performance_analysis performance_analysis_legend"><tr><td><strong>Legend:</strong></td>';
				html += '<td class="ok"><img src="/static/app/performance-analysis/images/blank.gif" alt="Infrastrucure OK" width="16" height="16" /></td><td>Infrastructure OK</td><td>&nbsp;</td>';
				html += '<td class="warning"><img src="/static/app/performance-analysis/images/blank.gif" alt="Infrastrucure Amber" width="16" height="16" /></td><td>Infrastructure Amber</td><td>&nbsp;</td>';
				html += '<td class="critical"><img src="/static/app/performance-analysis/images/blank.gif" alt="Infrastrucure Red" width="16" height="16" /></td><td>Infrastructure Red</td><td>&nbsp;</td>';
				html += '<td><img src="/static/app/performance-analysis/images/green.png" style="width:16px; height: 16px;" width="16" height="16" title="Application reported GREEN" /></td><td>Application Green</td><td>&nbsp;</td>';
				html += '<td><img src="/static/app/performance-analysis/images/amber.png" style="width:16px; height: 16px;" width="16" height="16" title="Application reported AMBER" /></td><td>Application Amber</td><td>&nbsp;</td>';
				html += '<td><img src="/static/app/performance-analysis/images/red.png" style="width:16px; height: 16px;" width="16" height="16" title="Application reported Red" /></td><td>Application Red</td><td>&nbsp;</td>';
				html += '<td class="downtime"><img src="/static/app/performance-analysis/images/blank.gif" alt="Down time" width="16" height="16" /></td><td>Down Time</td><td>&nbsp;</td></tr></table></td></table>';
			}
			
			return html;
		}
	}

	/* item class
		This class defines a row in the performance analysis graph.
		It will house the time buckets, item name etc.
		There will be multiple instances, but one per unique name
	*/
	class item{

		constructor(tname){
			this.name = tname;		// the item Name
			this.total = 0;			// the Total number of items
			this.buckets = [];		// The timebuckets
		}
	}

	/* timeBucket class
		This class is a collection of cells that will become the red/yellow/green/blue boxes in the analysis graph
	*/
	class  time_bucket {
		
		constructor(){
			this.start_time;
			this.end_time;
			this.num_failed = 0;
			this.num_amber = 0;
			this.num_red = 0;
			this.num_green = 0;
			this.total = 0;
			this.warning_threshold = 8;
			this.critical_threshold = 12;
			this.sum_value = 0.0;
			this.average_value = 0.0;
			this.showStatusAsText = true;
			this.category = ""
			// Status can be a float or string (12.20 or "GREEN")
			this.sum_status = 0.0;
			this.average_status = 0.0;
			//Colour Defaults:
			this.okColour = "#78B24A";
			this.warningColour = "#E0C135";
			this.criticalColour = "#DD0000";
			this.noDataColour = "#5EBFC6";
			this.isInDownTime = false;
			this.timeFormat = "h:mm A";
			this.oMoment = require('moment');
			this.OKIsLowOrHigh = "low";
		}
		

		set_colours(okColour, warningColour, criticalColour, noDataColour){
			this.okColour = okColour;
			this.warningColour = warningColour;
			this.criticalColour = criticalColour;
			this.noDataColour = noDataColour;
		}
		set_start_time(start){ 
			this.start_time = new Date(1970, 0, 1); // Epoch
			this.start_time.setSeconds(start);
	    }
		
		set_end_time (end){
			this.end_time = new Date(1970, 0, 1); // Epoch
			this.end_time.setSeconds(end);
	    };
		
		set_category(cat){
			this.category = cat
		}
		add_item(result_row,fields){
			var temp_status=0.0;
			var temp_value=0.0;
			try{
				//GREEN
				if (result_row[fields["status"]].toLowerCase()=="green") this.num_green++;
				if (result_row[fields["status"]].toLowerCase()=="ok") this.num_green++;
				if (result_row[fields["status"]].toLowerCase()=="pass") this.num_green++;
				if (result_row[fields["status"]].toLowerCase()=="low") this.num_green++;
				//AMBER
				if (result_row[fields["status"]].toLowerCase()=="amber") this.num_amber++;
                if (result_row[fields["status"]].toLowerCase()=="yellow") this.num_amber++;
				if (result_row[fields["status"]].toLowerCase()=="warn") this.num_amber++;
				if (result_row[fields["status"]].toLowerCase()=="warning") this.num_amber++;
				if (result_row[fields["status"]].toLowerCase()=="caution") this.num_amber++;
				//RED
				if (result_row[fields["status"]].toLowerCase()=="red") this.num_red++;
				if (result_row[fields["status"]].toLowerCase()=="critical") this.num_red++;
				if (result_row[fields["status"]].toLowerCase()=="severe") this.num_red++;
				if (result_row[fields["status"]].toLowerCase()=="fail") this.num_red++;
				
				// Add the passed, failed, update percentages etc
				temp_value=parseFloat(result_row[fields["value"]]);
				if (! isNaN(temp_value)) {
					this.sum_value += temp_value; // Increment total values for this bucket
					// Increment total values for this bucket
					this.total++;
				}
				this.average_value = (this.sum_value/ this.total).toFixed(2);
				
				temp_status=parseFloat(result_row[fields["status"]]);
				if (isNaN(temp_status)) { temp_status=0;}
				this.sum_status += temp_status;
				this.average_status = (this.sum_status/ this.total).toFixed(2); 
				
			} catch(err) {
				console.log("Error determining status / value.")
			}
		}
		
		getHTML(){
			var css_class= '';
			var html = '';
			var downtimeHTML = '';
			var colour = "";
			var td_image = "";
			var data_tooltip="";
			var time_period = "<b>Time Period:</b> " + this.oMoment(this.start_time).format(this.timeFormat) + " to " +  this.oMoment(this.end_time).format(this.timeFormat) + "</span><br />";
			if (this.total==0){
				td_image = "";
			}else if (!this.showStatusAsText && !isNaN(this.average_status)){
				td_image = "<span class=\"status_value\">" + this.average_status + "</span>";
				data_tooltip=time_period + "Average value: " + this.average_status;
			}else if (this.num_amber==0 && this.num_red==0 && this.num_green==0){
				td_image='<img src="/static/app/performance-analysis/images/blank.gif" style="width:16px; height: 16px;" width="16" height="16"/>';
				data_tooltip=time_period + '<b>Server:</b><br />Average: ' + this.average_value + '<br /><b>Application reported:</b><br />GREEN ' + this.num_green + ' time(s)<br />AMBER 0 time(s)<br />RED 0 time(s)';
			}else if (this.num_amber==0 && this.num_red==0 && this.num_green > 0){
				td_image='<img src="/static/app/performance-analysis/images/green.png" style="width:16px; height: 16px;" width="16" height="16"/>';
				data_tooltip=time_period + '<b>Server:</b><br />Average: ' + this.average_value + '<br /><b>Application reported:</b><br />GREEN ' + this.num_green + ' time(s)<br />AMBER 0 time(s)<br />RED 0 time(s)';
			}else if(this.num_amber >0 && this.num_red==0){ 
				td_image='<img src="/static/app/performance-analysis/images/amber.png" style="width:16px; height: 16px;" width="16" height="16"/>';
				data_tooltip=time_period + '<b>Server:</b><br />Average: ' + this.average_value + '<br /><b>Application reported:</b><br />GREEN ' + this.num_green + ' time(s)<br />AMBER ' + this.num_amber + ' time(s)<br />RED 0 time(s)';
			} else{
				td_image= '<img src="/static/app/performance-analysis/images/red.png" style="width:16px; height: 16px;" width="16" height="16"/>';
				data_tooltip=time_period + '<b>Server:</b><br />Average: ' + this.average_value + '<br /><b>Application reported:</b><br />GREEN ' + this.num_green + ' time(s)<br />AMBER ' + this.num_amber + ' time(s)<br />RED ' + this.num_red + ' time(s)';
			}
			
			// ---- Values (Background color) ----
			if(this.OKIsLowOrHigh=="low"){
				// OK is a low value
				if (this.average_value >= this.critical_threshold) { css_class += ' critical'; colour = this.criticalColour;}
				if (this.average_value < this.critical_threshold && this.average_value >= this.warning_threshold) { css_class += ' warning'; colour = this.warningColour;}
				if (this.average_value < this.warning_threshold) {css_class += ' ok'; colour = this.okColour;}
			}else{
				// OK is a high value
				if (this.average_value > this.warning_threshold) {css_class += ' ok'; colour = this.okColour;}
				if (this.average_value > this.critical_threshold && this.average_value <= this.warning_threshold) { css_class += ' warning'; colour = this.warningColour;}
				if (this.average_value <= this.critical_threshold) { css_class += ' critical'; colour = this.criticalColour;}
			
			}
			
			if (this.total==0) {css_class = " nodata"; colour=this.noDataColour;}
			if(this.isInDownTime) {
				css_class += ' downtime';
				data_tooltip += "<br>Downtime was in effect during this period.";
			}
			if (data_tooltip!="") {
				data_tooltip = " data-tooltip=\"" + data_tooltip + "\"";
			}
			html = '<td class="jds_ta_clickable ' + css_class + '" style="text-align:center;background-color: ' + colour + '!IMPORTANT" start_time="' + (this.start_time.getTime()/1000) + '" end_time="'+ (this.end_time.getTime()/1000) + '" category="name" value="' + this.category +'"' + data_tooltip + '>' + td_image + '</td>';
					
			return html;
		}
	}



	module.exports = { performance_analysis, item, time_bucket}
